src/Clapi/Valuespace.hs

```diff
@@ -348,22 +348,30 @@ processToRelayProviderDigest trpd vs =
       Valuespace tree' defs' (vsTyAssns vs) xrefs'
     return (updatedTypes, vs')
 
+validatePath :: MonadFail m => Valuespace -> Path -> Maybe (Set TpId) -> m ()
+validatePath vs p mTpids = do
+    def <- defForPath p vs
+    t <- note "Missing tree node" $ Tree.treeLookup p $ vsTree vs
+    claims <- validateRoseTreeNode def t mTpids
+    either (fail . show) return $ checkRefClaims vs $ Map.singleton p claims
+
 processToRelayClientDigest
   :: ContainerOps -> DataDigest -> Valuespace -> Map Path [Text]
 processToRelayClientDigest reords dd vs =
   let
     (updateErrs, tree') = Tree.updateTreeWithDigest reords dd (vsTree vs)
     vs' = vs {vsTree = tree'}
-    touchedPaths = Set.union (alKeysSet dd) (Map.keysSet reords)
-    touchedLiberties = Map.fromSet (flip getLiberty vs) touchedPaths
+    touched = ddTouched dd <> fmap (const Nothing) reords
+    touchedLiberties = Map.fromSet (flip getLiberty vs) $ Map.keysSet touched
```

Comment:
I _think_ this might be more efficient as `Map.mapWithKey (\p _ -> getLiberty p vs) touched` because `Map.mapWithKey` is a single O(n) operation, whereas `Map.fromSet` and `Map.keysSet` are both O(n), and I don't know if they combine efficiently...